functions/inputs/from.go

```diff
@@ -86,13 +91,293 @@ type FromProcedureSpec struct {
 
 	GroupingSet bool
 	OrderByTime bool
-	GroupMode   GroupMode
+	GroupMode   functions.GroupMode
 	GroupKeys   []string
 
 	AggregateSet    bool
 	AggregateMethod string
 }
 
+func (s FromProcedureSpec) PostPhysicalValidate(id plan.NodeID) error {
+	if !s.BoundsSet || (s.Bounds.Start.IsZero() && s.Bounds.Stop.IsZero()) {
+		return fmt.Errorf(`result '%s' is unbounded. Add a 'range' call to bound the query`, id)
+	}
+
+	return nil
+}
+
+// MergeFromRangeRule pushes a `range` into a `from`
+type MergeFromRangeRule struct{}
+
+// Name returns the name of the rule
+func (rule MergeFromRangeRule) Name() string {
+	return "MergeFromRangeRule"
+}
+
+// Pattern returns the pattern that matches `from -> range`
+func (rule MergeFromRangeRule) Pattern() plan.Pattern {
+	return plan.Pat(transformations.RangeKind, plan.Pat(FromKind))
+}
+
+// Rewrite attempts to rewrite a `from -> range` into a `FromRange`
+func (rule MergeFromRangeRule) Rewrite(node plan.PlanNode) (plan.PlanNode, bool, error) {
+	from := node.Predecessors()[0]
+	fromSpec := from.ProcedureSpec().(*FromProcedureSpec)
```

Comment:
It would be nice if we did matching rules on the actual types and the kind was only used for JSON encoding. I always feel a bit uncomfortable with encoding the kind into logic like this, but it's everywhere else so this comment is by no means a suggestion to change this.

I would like to revisit it in the future though.

Reply:
I'm open to using something else.  Ultimately, we want our patterns to be able to not just recognize a procedure of a particular type, but some class of procedures within a taxonomy (like any source that can have a filter pushed into it, any aggregate function, etc).